src/core/ext/transport/chttp2/transport/context_list.h

```diff
@@ -32,18 +36,59 @@ namespace grpc_core {
 class ContextList {
```

Comment:
Next question - and I don't have an answer.

Do we need this class or would just having `ContextListEntry` be enough?

Argument against having it:

Append is just an `emplace_back` if `ContextListEntry` has the right constructor.
IsEmpty is trivial.
MakeNewContextList is redundant (just instantiate a `std::vector<ContextListEntry>`)
`ForEachExecuteCallback` is just a really complicated way of writing `for (const auto& ent : entries) { ... }`

Reply:
Or can we refactor in a separate PR? I guest we can just remove ContextList and replace it with vector<ContextListEntry> with a little more work, but given we have a customer who wants to enable Dapper asap we don't need to rush this in this PR.

Reply:
So a few things:
- the code you submit usually lasts longer than you'd like
- there's always another urgent customer
- added complexity tends to breed more complexity

Taken together, if there's a really quick change to wipe out an entire module I think we should take that... It doesn't significantly delay things now, and it speeds us up on the next thing.

Reply:
Sure, I think it's ultimately you and Vignesh's call. The reason I proposed to separate them out is that I expect we need the Dapper support enabled for the customer asap within days, then it's worth not mixing any changes.

But can we at least remove the two additional args here and make this a pure refactor PR? I think we can add them in the "Dapper" PR and make it one, I expect this will make (unlikely) rollback of this refactor easier too.

Reply:
I think it's better to keep ContextList as a thin wrapper class instead of exposing the vector directly.

Both Append() and Execute() are non-trivial and the underlying vector is only supposed to be updated in specific ways, so I think they should remain in one place and hidden. I assume this simply requires moving a few lines of code back.

Reply:
The pre-existing APIs seem a bit non intuitive. I unrolled the Append() logic directly into the chttp2/writing.cc  file because its the only place its used. 

By dealing with the vector directly, other internal EventEngine implementations can directly iterate over the ContextListEntries instead of exposing another API to the wrapper ContextList class to iterate and process them.

@ctiller do you agree ?

Reply:
It's your call, but I think this is a clear case to hide internals. There are only two operations that are expected and current APIs look reasonable to me. We don't have unit tests which I think we need, but if we had them it might be another good reason to hide unnecessary internals and only test the APIs. Append() being called from one place seems less relevant given this is more about the structure, and I don't think we have other places that need to iterate over them now.